src/yandex_cloud_ml_sdk/_datasets/dataset.py

```diff
@@ -136,6 +142,68 @@ async def _delete(
 
         logger.info("Dataset %s successfully deleted", self.id)
 
+    @safe_on_delete
+    async def _download(
+        self,
+        *,
+        download_path: PathLike,
+        timeout: float = 60,
+    ) -> tuple[Path, ...]:
+        logger.debug("Downloading dataset %s", self.id)
+
+        base_path = coerce_path(download_path)
+        if not base_path.exists():
+            raise ValueError(f"{base_path} does not exist")
+
+        if not base_path.is_dir():
+            raise ValueError(f"{base_path} is not a directory")
+
+        if os.listdir(base_path):
+            raise ValueError(f"{base_path} is not empty")
+
+        return await asyncio.wait_for(self.__download_impl(
+            base_path=base_path,
+        ), timeout)
+
+    async def __download_impl(
+        self,
+        *,
+        base_path: Path
+    ) -> tuple[Path, ...]:
+        urls = await self._get_download_urls()
+        async with self._client.httpx() as client:
+            coroutines = [
+                self.__download_file(base_path / key, url, client) for key, url in urls
+            ]
+
+            await asyncio.gather(*coroutines)
+
+        return tuple(base_path / key for key, _ in urls)
+
+    async def __download_file(
+        self,
+        path: Path,
+        url: str,
+        client: httpx.AsyncClient,
+    ) -> None:
+        async with aiofiles.open(path, "wb") as file:
+            async for chunk in self.__read_from_url(url, client):
+                await file.write(chunk)
+            await file.flush()
```

Comment:
I would expect aiofiles will do flush by itself when it will exit the context manager